internal/controller/pipelinerollout_progressive.go

```diff
@@ -155,5 +165,86 @@ func (r *PipelineRolloutReconciler) updatePipelineLabel(
 		return err
 	}
 	return nil
+}
+
+func (r *PipelineRolloutReconciler) cleanUpPipelines(
+	ctx context.Context,
+	pipelineRollout *apiv1.PipelineRollout,
+) error {
+	recyclablePipelines, err := r.getRecyclablePipelines(ctx, pipelineRollout)
+	if err != nil {
+		return err
+	}
+
+	for _, recyclablePipeline := range recyclablePipelines {
+		err = r.processRecyclablePipelineStatus(ctx, recyclablePipeline)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// getRecyclablePipelines retrieves all the recyclable pipelines managed by the given
+// pipelineRollout through the `recyclable` label.
+func (r *PipelineRolloutReconciler) getRecyclablePipelines(
+	ctx context.Context,
+	pipelineRollout *apiv1.PipelineRollout,
+) ([]*kubernetes.GenericObject, error) {
+	return kubernetes.ListCR(
+		ctx, r.restConfig, common.NumaflowAPIGroup, common.NumaflowAPIVersion, "pipelines",
+		pipelineRollout.Namespace, fmt.Sprintf(
+			"%s=%s,%s=%s", common.LabelKeyPipelineRolloutForPipeline, pipelineRollout.Name,
+			common.LabelKeyUpgradeState, common.LabelValueUpgradeRecyclable,
+		), "")
+}
+
+func (r *PipelineRolloutReconciler) processRecyclablePipelineStatus(
+	ctx context.Context,
+	pipelineDef *kubernetes.GenericObject,
+) error {
+	pipelineStatus, err := parsePipelineStatus(pipelineDef)
+	if err != nil {
+		return fmt.Errorf("failed to parse Pipeline Status from pipeline CR: %+v, %v", pipelineDef, err)
+	}
+	pipelinePhase := pipelineStatus.Phase
+
+	// Only delete fully drained pipelines
+	if pipelinePhase == numaflowv1.PipelinePhasePaused {
+		// check if `pipelineStatus.DrainedOnPause` is true
+		if pipelineStatus.DrainedOnPause {
+			err = kubernetes.DeleteCR(ctx, r.restConfig, pipelineDef, "pipelines")
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func parsePipelineStatus(obj *kubernetes.GenericObject) (numaflowv1.PipelineStatus, error) {
+	if obj == nil || len(obj.Status.Raw) == 0 {
+		return numaflowv1.PipelineStatus{}, nil
+	}
+
+	var status numaflowv1.PipelineStatus
+	err := json.Unmarshal(obj.Status.Raw, &status)
+	if err != nil {
+		return numaflowv1.PipelineStatus{}, err
+	}
 
+	return status, nil
+}
+
+func isPipelineReady(status numaflowv1.Status) bool {
+	if len(status.Conditions) == 0 {
+		return false
+	}
+	for _, c := range status.Conditions {
+		if c.Status != metav1.ConditionTrue {
+			return false
+		}
+	}
+	return true
 }
```

Comment:
nit: do you mind if we move `parsePipelineStatus()` and `isPipelineReady()` into pipelinerollout_controller.go since those will be potentially useful outside of progressive strategy?